hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/SqmAnyValuedSimplePath.java

```diff
@@ -44,6 +44,7 @@ public SqmAnyValuedSimplePath(
 
 	@Override
 	public SqmAnyValuedSimplePath<T> copy(SqmCopyContext context) {
+		final SqmPath<?> lhsCopy = getLhs().copy( context );
```

Comment:
If `this` path was already copied before, copying the `lhs` here should have no effect, hence why it was only done when the `this` path was not yet copied. Can you please explain a bit what is happening here and why this makes a difference?

Reply:
OFC: when `getLhs()` was a treated path (e.g. `SqmTreatedSimplePath`), the `copy()` method of that triggered another copy of `this` while the first one was still happening - this is because `this` would be in the `reusablePaths` of the treated path object. The second copy of `this` triggered by `getLhs().copy()` (the treated path) would then be registered but when that finished we would still be inside the original `copy` method of `this`. This is what caused the exception *Already registered a copy*.

Moving the `getLhs().copy()` method up, before the check for existing copies, solves this issue: when the treated path copies `this` now the copy is registered early and we don't try to create another because we will find the existing one.

Reply:
Sounds like we should rather defer the copying of the `reusablePaths` or maybe even skip that altogether, since the reusable paths will be visited eventually anyway throughout the copying.

Reply:
I did try skipping that and it looked like it would solve the issue, but wouldn't the `getLhs()` copy then be missing the `reusablePaths`? Also, deferring would be hard because we would need to defer the `getLhs().copyTo()` after `this` copy has finished but we have no way of knowing that from the treated path.